src/components/container.js

```diff
@@ -179,13 +186,167 @@ export default class Container extends Component {
 }
 
 export function createContainer(
-  Store,
+  StoreOrOptions = {},
   { onInit = noop, onUpdate = noop, onCleanup = noop, displayName = '' } = {}
 ) {
-  return class extends Container {
-    static storeType = Store;
-    static displayName =
-      displayName || `Container(${Store.key.split('__')[0]})`;
-    static hooks = { onInit, onUpdate, onCleanup };
+  if ('key' in StoreOrOptions) {
+    const Store = StoreOrOptions;
+    const dn = displayName || `Container(${Store.key.split('__')[0]})`;
+
+    return class extends Container {
+      static storeType = Store;
+      static displayName = dn;
+      static hooks = { onInit, onUpdate, onCleanup };
+    };
+
+    // eslint-disable-next-line no-unreachable
+    return createFunctionContainer({
+      displayName: dn,
+      // compat fields
+      override: {
+        Store,
+        handlers: {
+          ...(onInit !== noop && { onInit: () => onInit() }),
+          ...(onCleanup !== noop && { onDestroy: () => onCleanup() }),
+          ...(onUpdate !== noop && { onContainerUpdate: () => onUpdate() }),
+        },
+      },
+    });
+  }
+
+  return createFunctionContainer(StoreOrOptions);
+}
+
+function useRegistry(scope, isGlobal, { globalRegistry }) {
+  return useMemo(() => {
+    const isLocal = !scope && !isGlobal;
+    return isLocal ? new StoreRegistry('__local__') : globalRegistry;
+  }, [scope, isGlobal, globalRegistry]);
+}
+
+function useContainedStore(scope, registry, props, override) {
+  // Store contained scopes in a map, but throwing it away on scope change
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const containedStores = useMemo(() => new Map(), [scope]);
+
+  // Store props in a ref to avoid re-binding actions when they change
+  // If devs want to update consumers on prop change, they can put them in store
```

Comment:
Sorry, that comment is misleading, just updated it. The reason why we store props in a ref is to avoid all consumers to re-render (as the context value depends on `getContainedStore`). Instead. we use a useEffect in the component to manually trigger `onContainerUpdate` explicitly and pass latest props via a getter.